### PR TITLE
Local glean dev

### PR DESCRIPTION
**Pull request by badboy**
PR over at android-components: https://github.com/mozilla-mobile/android-components/pull/5129

### CODE REVIEW

File: settings.gradle [hunk 1]
```diff
@@ -11,6 +11,7 @@ buildscript {
         jcenter()
     }
 }
+rootProject.name = "glean-core"
```
Comment (Dexterp37) [inline settings.gradle]: Does this have any impact on consuming the library outside of the Glean repo? I remember messing with this property a while back and breaking things.

Comment (badboy) [reply settings.gradle]: That is indeed something I should check.

Comment (badboy) [reply settings.gradle]: (I have an abandoned commit around which just renames all glean-core mentions to glean, we could also just use that)

File: docs/dev/howtos/development-with-android-components.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+There's a big gotcha with library substitutions: the Gradle build computes lazily, and AARs don't include their transitive dependencies' JNI libraries.
+This means that in `android-components`, `./gradlew :service-glean:assembleDebug` **does not** invoke `:glean-core:cargoBuild`,
+even though `:service-glean` depends on the substitution for `:glean-core` and even if the inputs to Cargo have changed!
+It's the final consumer of the `:service-glean` project (or publication) that will incorporate the JNI libraries.
+
+In practice that means _you should always be targeting something that produces an APK_: a test or a sample module.
+Then you should find that the `cargoBuild` tasks are invoked as you expect.
+
+`./gradlew :samples-glean:connectedAndroidTest` should work.  
+Other tests like `:glean-core:testDebugUnitTest` or `:support-sync-telemetry:testDebugUnitTest` will currently fail, because the JNI libraries are not included.
```
Comment (mdboom) [inline docs/dev/howtos/development-with-android-components.md]: I find this section a bit confusing since it doesn't specify which project (a-c or glean.rs) that these tests are run in, and it mixes targets in `a-c` (`samples-glean`) and glean (`glean-core`).

I think the general idea here is that you don't have to re-run anything the the glean checkout, just that you have to run a apk-producing target in the `a-c` checkout.  Is that correct?

File: settings.gradle (hunk 1 above)
Comment (Dexterp37) [reply settings.gradle]: (that would be cool!)

File: docs/dev/howtos/development-with-android-components.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+
+## Notes
+
+This document is based on the equivalent documentation for application-services:
+[Development with the Reference Browser](https://github.com/mozilla/application-services/blob/master/docs/howtos/working-with-reference-browser.md)
+
+1. Transitive substitutions (as shown above) work but require newer Gradle versions (4.10+).
+1. Android Studio happily imports substitutions, but it doesn't appear to always do the right thing with transitive substitutions.  
+   Best to keep substitutions in the final project (i.e., in `reference-browser/settings.gradle`) or to duplicate them in all transitive links.
```
Comment (mdboom) [inline docs/dev/howtos/development-with-android-components.md]: "final project" meaning "application" here?

File: docs/dev/howtos/development-with-android-components.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+1. Be aware that the project list can get very large!  At this time, there's no way to filter the project list.
```
Comment (mdboom) [inline docs/dev/howtos/development-with-android-components.md]: What is the "project list" in this context?

File: docs/dev/howtos/locally-published-components-in-fenix.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
+# Using locally-published Glean in Fenix
+
+> Note: This is a bit tedious, and you might like to try the substitution-based > approach documented in
+> [Working on unreleased Glean code in android-components](./development-with-android-components.md).
+> That approach is still fairly new, and the local-publishing approach in this document is necessary if it fails.
```
Comment (mdboom) [inline docs/dev/howtos/locally-published-components-in-fenix.md]:
```suggestion
> That approach is still fairly new, and the local-publishing approach in this document is only necessary if it fails.
```

File: docs/dev/howtos/locally-published-components-in-fenix.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+git clone https://github.com/mozilla-mobile/android-components
+git clone https://github.com/mozilla-mobile/fenix/
+```
+
+## Local publishing
+
+
+1. Inside the `glean` repository root:
+    1. In [`.buildconfig.yml`](glean-yaml), change
+       `libraryVersion` to end in `-TESTING$N` <sup><a href="#note1">1</a></sup>,
```
Comment (mdboom) [inline docs/dev/howtos/locally-published-components-in-fenix.md]: I think mdbook supports the [markdown footnote syntax](https://www.markdownguide.org/extended-syntax/#footnotes).  (It's used elsewhere in our docs).

File: docs/dev/howtos/locally-published-components-in-fenix.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
+# Using locally-published Glean in Fenix
+
+> Note: This is a bit tedious, and you might like to try the substitution-based > approach documented in
```
Comment (mdboom) [inline docs/dev/howtos/locally-published-components-in-fenix.md]:
```suggestion
> Note: This is a bit tedious, and you might like to try the substitution-based approach documented in
```

File: docs/dev/howtos/locally-published-components-in-fenix.md [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+    4. Inside the android-component's `local.properties` file, ensure
+       `substitutions.glean.dir` is *NOT* set.
+
+    5. Run `./gradlew publishToMavenLocal`.
+
+3. Inside the `fenix` repository root:
+    1. Inside [`build.gradle`](fenix-build-gradle-1), add
+       `mavenLocal()` inside `allprojects { repositories { <here> } }`.
+        1. If you added a new project to the megazord (e.g. you went through the
```
Comment (mdboom) [inline docs/dev/howtos/locally-published-components-in-fenix.md]: Does megazording apply to glean-core?

**Comment by badboy** [comment]
Guess I need to finally mix my linkcheck setup locally

File: docs/dev/howtos/locally-published-components-in-fenix.md (hunk 8 above)
Comment (badboy) [reply docs/dev/howtos/locally-published-components-in-fenix.md]: Woops, thought I removed that line

File: docs/dev/howtos/development-with-android-components.md (hunk 2 above)
Comment (badboy) [reply docs/dev/howtos/development-with-android-components.md]: o_O Woops, the mix of samples-glean and glean-core is a copy-pasta mistake. This should all be run inside a-c. I'll fix that and make it more clear it's all in a-c.

File: docs/dev/howtos/development-with-android-components.md (hunk 3 above)
Comment (badboy) [reply docs/dev/howtos/development-with-android-components.md]: yes, let me rephrase that as well

File: docs/dev/howtos/development-with-android-components.md (hunk 4 above)
Comment (badboy) [reply docs/dev/howtos/development-with-android-components.md]: Can be removed, turns out the `:glean` projects list isn't included in anything visible.

File: docs/dev/howtos/locally-published-components-in-fenix.md (hunk 6 above)
Comment (badboy) [reply docs/dev/howtos/locally-published-components-in-fenix.md]: True, let me try that

File: settings.gradle (hunk 1 above)
Comment (Dexterp37) [reply settings.gradle]: Did you check this?
